drivers/adc/adc_ads1x1x.c

```diff
@@ -0,0 +1,662 @@
+		adc_context_complete(&data->ctx, rc);
+		return rc;
+	}
+
+	int16_t buf;
+
+	rc = ads1x1x_read_reg(dev, ADS1X1X_REG_CONV, &buf);
+	if (rc != 0) {
+		adc_context_complete(&data->ctx, rc);
+		return rc;
+	}
+	/* The ads101x stores it's 12b data in the upper part
+	 * while the ads111x uses all 16b in the register, so
+	 * shift down. Data is also signed, so perform
+	 * division rather than shifting
+	 */
+	*data->buffer++ = buf / (1 << (15 - config->resolution));
+
+	adc_context_on_sampling_done(&data->ctx, dev);
+
+	return rc;
+}
+
+static int ads1x1x_read(const struct device *dev, const struct adc_sequence *sequence)
+{
+	return ads1x1x_adc_read_async(dev, sequence, NULL);
+}
+
+static void ads1x1x_acquisition_thread(const struct device *dev)
+{
+	while (true) {
+		ads1x1x_adc_perform_read(dev);
+	}
+}
+
+static int ads1x1x_init(const struct device *dev)
+{
+	const struct ads1x1x_config *config = dev->config;
+	struct ads1x1x_data *data = dev->data;
+
+	data->dev = dev;
+
+	k_sem_init(&data->acq_sem, 0, 1);
+
+	if (!device_is_ready(config->bus.bus)) {
+		LOG_ERR("I2C bus %s not ready", config->bus.bus->name);
+		return -ENODEV;
+	}
+
+	const k_tid_t tid =
+		k_thread_create(&data->thread, data->stack, K_THREAD_STACK_SIZEOF(data->stack),
+				(k_thread_entry_t)ads1x1x_acquisition_thread, (void *)dev, NULL,
+				NULL, CONFIG_ADC_ADS1X1X_ACQUISITION_THREAD_PRIO, 0, K_NO_WAIT);
+	k_thread_name_set(tid, "adc_ads1x1x");
+
+	adc_context_unlock_unconditionally(&data->ctx);
+
+	return 0;
+}
+
+static const struct adc_driver_api ads1x1x_api = {
+	.channel_setup = ads1x1x_channel_setup,
+	.read = ads1x1x_read,
+	.ref_internal = 4096,
+#ifdef CONFIG_ADC_ASYNC
+	.read_async = ads1x1x_adc_read_async,
+#endif
+};
+
+#define DT_INST_ADS1X1X(inst, t) DT_INST(inst, ti_ads##t)
+
+#define ADS1X1X_INIT(t, n, odr_delay_ms, res, mux, pgab) \
+	static const struct ads1x1x_config ads##t##_config_##n = { \
+		.bus = I2C_DT_SPEC_GET(DT_INST_ADS1X1X(n, t)), \
+		.odr_delay = odr_delay_ms, \
+		.resolution = res, \
+		.multiplexer = mux, \
+		.pga = pgab, \
+	}; \
+	static struct ads1x1x_data ads##t##_data_##n = { \
+		ADC_CONTEXT_INIT_LOCK(ads##t##_data_##n, ctx), \
+		ADC_CONTEXT_INIT_TIMER(ads##t##_data_##n, ctx), \
+		ADC_CONTEXT_INIT_SYNC(ads##t##_data_##n, ctx), \
+	}; \
+	DEVICE_DT_DEFINE(DT_INST_ADS1X1X(n, t), ads1x1x_init, NULL, &ads##t##_data_##n, \
+			 &ads##t##_config_##n, POST_KERNEL, CONFIG_ADC_ADS1X1X_INIT_PRIORITY, \
+			 &ads1x1x_api);
+
+#define ADS1X1X_FOREACH_STATUS_OKAY(compat, fn) \
+	COND_CODE_1(DT_HAS_COMPAT_STATUS_OKAY(compat), \
+		    (UTIL_CAT(DT_FOREACH_OKAY_INST_, compat)(fn)), ())
+
+/* The ADS111X provides 16 bits of data in binary two's complment format
+ * A positive full-scale (+FS) input produces an output code of 7FFFh and a
+ * negative full-scale (–FS) input produces an output code of 8000h. Single
+ * ended signal measurements only only use the positive code range from
+ * 0000h to 7FFFh
+ */
+#define ADS111X_RESOLUTION 15
```

Comment:
The result is actually a 16-bit value (yes, in two's complement format, but it uses all 16 bits), also the datasheet mentions 16-bit resolution everywhere, so I think users will be confused if the driver will force them to specify 15 bits in the sequence definition (for 16 bits, they will get an error).

Reply:
yeah, but according to the definition of resolution from here: https://docs.zephyrproject.org/latest/reference/peripherals/adc.html#c.adc_sequence.resolution

>ADC resolution. For single-ended channels the sample values are from range: 0 .. 2^resolution - 1, for differential ones: 2^(resolution-1) .. 2^(resolution-1) - 1.

This would still be 15 bits of the user is not in differential mode. I should probably update it so, if the user is in differential mode, they should configure for 16 bits according to the definition and it should check accordingly.

Reply:
I'd have to think more about how to implement this. It's not in the last push yet.

Reply:
@anangl What I'm thinking here is to add a flag in the driver data that will record if it's only channel is configured for differential or single ended when `adc_channel_setup` is called. Then when the user calls `adc_read`, then it will check if `resolution` is ==15 or ==16 based on the flag that was recorded in the channel setup....